Fix various annotation string printf format issues

markcmiller86 opened the pull request:
### Description

Resolves #18609 
Resolves #18922
Resolves #19385

I added a slew of tests that do the following...

- Test all the different annotation macros in both 2D and 3D text
- Test multiple macros in the same string
- Test various printf-style formatting of macros
- Test tafile macros

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* [x] Bug fix~~
* ~~[ ] New feature~~
* ~~[ ] Documentation update~~
* ~~[ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

<!-- Please describe the tests you've added or any tests that already cover this change. Include relevant information, such as which operating system you tested on. -->

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [x] I have commented my code where applicable.~~
- [x] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- ~~[ ] I have added debugging support to my changes.~~
- [x] I have added tests that prove my fix is effective or that my feature works.~~
- [x] I have confirmed new and existing unit tests pass locally with my changes.~~
- [x] I have added new baselines for any new tests to the repo.~~
- [x] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

markcmiller86 committed:
fix time format string

markcmiller86 committed:
upd relnotes

markcmiller86 commented:
@biagas I think the correct fix on `develop` is to remove the **Time format** text box from the GUI. It is basically superfulous now as the format string can be part of the annotation string itself. Thoughts?

biagas commented on src/avt/VisWindow/Colleagues/avtTimeSliderColleague.C:
```diff
@@ -712,8 +713,46 @@ avtTimeSliderColleague::SetText(const char *formatString, const char *timeFormat
+        textString = CreateAnnotationString(combined.c_str());
+    }
+    else
+    {
```
This all seems unnecessarily complicated. Having the printf-type format string be part of the source(text) string  (for me) hard to parse. If the ability to set a simple `%g` or similar type string goes away, are users going to have difficulty parsing the DEFAULT string of `$time%g` that will need to appear in the text field? Or know how to modify it correctly?


I tried  reading CreateAnnotationString, and came up wondering why we aren't passing the printf-type format string directly to the method and using an sprintf there to create the final string to be displayed, instead of combining it with the source string and performing such complicated parsing.

cyrush replied:
c++'s fmt lib deals with complex formatting with more embedded markup.

Here is a general example:

https://github.com/LLNL/conduit/blob/0d8db6d00441d97a095f817bd416c3e04eec153c/src/tests/conduit/t_conduit_utils.cpp#L720

Instead of implement a lot of logic ourselves, we might want to use format to help with complex cases.

markcmiller86 replied:
Please have a look at this [documentation](https://github.com/visit-dav/visit/blob/0861518abf80bf16a967495fd12526623d437adb/src/avt/VisWindow/Colleagues/avtAnnotationWithTextColleague.C#L229) and this [comment](https://github.com/visit-dav/visit/blob/0861518abf80bf16a967495fd12526623d437adb/src/avt/VisWindow/Colleagues/avtAnnotationWithTextColleague.C#L229) in the source code, in particular paying attention to how *default* format is handled. In addition lemme know how printf-format can be handled in cases where we specify multiple, same or different `$macros` in the same annotation string, as well as the ability to handle all the different annotation macros that are now supported, as well as the ability to easily add new ones in the future (which was extremely useful when @ndimensionalspace wanted to take annotation strings from files), and lemme know if you still think it is "unnecesarily complicated". BTW, you approved all that work in #17811 two years ago.

biagas replied:
If the process of creating the string that gets passed to CreateAnnotationString (and that method itself) works as intended, why do we need to also allow user to specify the printf-format in the text string like this `Time=$time%6.4`?

You also asked about doing away with the field that only specifies the printf-format altogether and forcing it to be part of the text string, which is also what I was attempting to question in my previous comment.

biagas replied:
Mark, thanks for the reminder of all the different ways that text can be specified  in an Annotation Object (eg multiple $ macros plus printf-format strings). My thinking cap was unfortunately the simplest cases.

Did you ever add tests for these more complex use-cases? If not I will create  ticket.

markcmiller86 replied:
> Did you ever add tests for these more complex use-cases? If not I will create ticket.

I did not but wanted do and I honestly think now is best time. So, I will work to add some of that here in this PR.

markcmiller86 replied:
Ok, I've added tests. Now requesting re-review

markcmiller86 replied:
BTW...the tests ran on macOS. Could be different from poodle. If so, will re-base after nightlies run.

JustinPrivitera commented on src/avt/VisWindow/Colleagues/avtTimeSliderColleague.C:
```diff
@@ -712,8 +713,46 @@ avtTimeSliderColleague::SetText(const char *formatString, const char *timeFormat
+
+    // 
+    // We first check for possible printf-style format string within the
+    // textFormatString itself. If we find something string there, we just
```
find something string? Maybe typo? Or am I not reading the sentence right?

markcmiller86 replied:
Thanks fixed now.

JustinPrivitera reviewed (commented):
Curious about the resolution to Kathleen's comments.

markcmiller86 committed:
add baselines

markcmiller86 committed:
add test

markcmiller86 committed:
fix docs

markcmiller86 committed:
fix typo in comment

markcmiller86 committed:
upd %-style formatting

markcmiller86 committed:
Merge branch '3.4RC' into bug-mcm86-05nov24-annot-str-printf

markcmiller86 commented:
@JustinPrivitera and @biagas can you please review now.

markcmiller86 committed:
add more verbiage

markcmiller86 committed:
Merge branch 'bug-mcm86-05nov24-annot-str-printf' of github.com:visit-dav/visit into bug-mcm86-05nov24-annot-str-printf

JustinPrivitera commented on src/doc/python_scripting/quickrecipes.rst:
```diff
@@ -678,6 +678,14 @@ Note that this approach can also be used to insert images of graphs, plots, port
 
    Image annotation used to incorporate LLNL logo
 
+Deleting an Annotation Object
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To delete any annotation object created via ``CreateAnnotationObject(...)``, just use the returned object's ``Delete()`` method.
+For example, in the preceding code to create an ``"Image"`` annotation object, the returned object is stored in the ``image`` python variable.
+To later delete that object, use ``image.Delete()``.
+If for some reason, the object handle is *lost*, maybe because the original python variable holding it was overwritten, it can be re-acquired using ``GetAnnotationObject(<objname>)`` where ``<objname>`` is a string identifying the annotation object.
+In addition, ``GetAnnotationObjectNames()`` can be used to return a list of names of the currently defined annotation objects.
+
```
This is great

JustinPrivitera commented on test/baseline/rendering/annot_macros/text3d_fulldbname.png:
This will change when it is running as part of the test suite, right?

markcmiller86 replied:
Oh, yep. Good point. I will **definitely** have to rebase that one for sure.

JustinPrivitera commented on src/test/tests/rendering/annot_macros.py:
```diff
@@ -0,0 +1,186 @@
+#
+# 71 string values to write to a tafile, value @ index 13 is "Monument"
+#
+svals = ["Ability", "Zealous", "Question", "Yarn", "Wisdom", "Venture", "Umbrella",
+    "Triumph", "Sincere", "Revelation", "Porcelain", "Optimism", "Novelty", "Monument",
+    "Latitude", "Kinship", "Journey", "Illumination", "Harvest", "Guidance", "Fidelity",
+    "Elixir", "Duration", "Curiosity", "Blueprint", "Aspire", "Artifact", "Validation",
+    "Unfold", "Theory", "Sanctuary", "Retreat", "Perspective", "Narrative", "Mastery",
+    "Longevity", "Keystone", "Intrigue", "Heritage", "Generation", "Fluid", "Enigma",
+    "Delight", "Craft", "Beacon", "Anonymous", "Achieve", "Whisper", "Vigilant", "Unity",
+    "Tranquil", "Symbolic", "Resilience", "Quest", "Purity", "Overture", "Mirth",
+    "Luminous", "Kindred", "Junction", "Insight", "Horizon", "Gradient", "Flourish",
+    "Essence", "Destiny", "Cycle", "Brevity", "Awakening", "Altruism", "Zenith"]
```
I am confused; what are all of these for?

markcmiller86 replied:
Well, these are to test the `$TtafileX` (and friends) style macros. These macros work by storing lines of data in a `.txt` file, one for each timestate of the database. You can [read more about them](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/using_visit/MakingItPretty/Annotations.html#named-database-values-in-text-annotations)

I had asked ChatGpt to create a random collection of 71 English language words to test `stafile1`. But, now that I think about it, those aren't all that useful in testing this feature. For example, you have to count in 13 states to find that the right string to have rendered after having called `TimeSliderSetState(13)` is `"Monument"`.

Should I adjust?

JustinPrivitera replied:
I don't think it's a big deal; if you want to trim the list down a bit maybe that would be better. But if not no worries.

markcmiller86 committed:
fix topodim/spatialdim doc

markcmiller86 committed:
use modern format

markcmiller86 committed:
Adjust testing logic to modern structure

markcmiller86 commented:
Ok, I've made some adjustments in the testing part of this PR. First, I switched to the more modern form of structuring a `.py` test file including using functions for sections of tests and using `TestAutoName()`.

I don't know why I didn't think of this before but instead of having 3 different images for each annotation macro rendered as 2D text, 3D text and time slider text, I just put all into the same image. That reduces the number of baselines a lot. Next, I also removed baselines involving macros that the database had no data for (e.g. `lod`, `xunits`, etc.)

Finally, as @JustinPrivitera observed, I adjusted the logic to create input files for `$TtafileI.txt` cases and made the data more "human readable" in the sense that the values can be more easily connected to their associated database index.

So, a lot of changes to testing but I think its much improved.

markcmiller86 committed:
fix a test

JustinPrivitera reviewed (approved):
The tests look great.
